### DISCUSSION TITLE
websocket error

### DISCUSSION THREAD

**Discussion by KarianneM** [getgauge/taiko#1879] [Development]
Hello,

We have a spec that searches and downloads files
After some repeats of browsing steps ( search and download file) I encounter these errors:

TypeError: Cannot read property 'webSocketDebuggerUrl' of undefined
at Chrome._fetchDebuggerURL (D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:182:31)
at runMicrotasks (<anonymous>)
at processTicksAndRejections (internal/process/task_queues.js:93:5)
at async Chrome._start (D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:139:25)

it seems like an earlier error can't be described. 

or 

Error: WebSocket is not open: readyState 2 (CLOSING)
at sendAfterClose (D:\clicking_tool\node_modules\ws\lib\websocket.js:754:17)
at WebSocket.send (D:\clicking_tool\node_modules\ws\lib\websocket.js:345:7)
at Chrome._enqueueCommand (D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:281:18)
at D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:87:22

Chrome Versie 88.0.4324.182 
Taiko v 1.1.4

We've created a new similar project and there the error is: "no connection can be made because the target machine actively refuses it"

Taiko v 1.2.0

The code we use is 

      await click('xxxxx');  
      await goto(link,5000);
      await waitFor('2000')
      await click("Search", above ("xxxxx"));

it can go correct for 50 times or only once before erroring out. Did anyone get this too or does anyone know how to catch the error to continue the other steps?

**Comment by NivedhaSenthil** [comment]
`no connection can be made because the target machine actively refuses it` can you please the full stacktrace if available ?

**Reply by KarianneM** [reply]
Hi Nivedha,

Here it is: (I've removed the ip address and the name of the step)

Error: WebSocket is not open: readyState 2 (CLOSING)
at sendAfterClose (D:\clicking_tool\node_modules\ws\lib\websocket.js:754:17)
at WebSocket.send (D:\clicking_tool\node_modules\ws\lib\websocket.js:345:7)
at Chrome._enqueueCommand (D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:281:18)
at D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:87:22
at new Promise (<anonymous>)
at Chrome.send (D:\clicking_tool\node_modules\chrome-remote-interface\lib\chrome.js:86:20)
at handler (D:\clicking_tool\node_modules\chrome-remote-interface\lib\api.js:32:23)
at D:\clicking_tool\node_modules\taiko\lib\connection.js:31:33
at new Promise (<anonymous>)
at Proxy.<anonymous> (D:\clicking_tool\node_modules\taiko\lib\connection.js:22:24)

        Error Message: Runner is not Alive
        Stacktrace: 
        

        Failed Step: Search for Document
        Specification: specs\0110_search.spec:118
        Error Message: transport is closing
        Stacktrace: 
        

    Error Message: Runner is not Alive
    Stacktrace: 
    

  Error Message: Runner is not Alive
  Stacktrace: 
  


  Error Message: Runner is not Alive
  Stacktrace: 
  
Successfully generated html-report to => reports\html-report\2021-02-19 15.45.18\index.html

Failed to kill Runner: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp xxxxx: connectex: No connection could be made because the target machine actively refused it."
Specifications: 4 executed      3 passed        1 failed        265 skipped
Scenarios:      0 executed      0 passed        1 failed        25 skipped

**Comment by KarianneM** [comment]
Is this the right information?

**Comment by KarianneM** [comment]
In the latest failed tests of the pull requests similar errors occur, could that be related?

https://github.com/getgauge/taiko/pull/1886/checks?check_run_id=1967619755

**Reply by NivedhaSenthil** [reply]
This may not be related... `no connection can be made because the target machine actively refuses it` this seems to be happening in Gauge due to some failure in the test script.. will be helpful for debugging if you can provide simple replica for us to try

**Reply by KarianneM** [reply]
Thanks a lot for thinking with me, the site in an internal one so I can't put the url

the basis is like this:

```
step("Goto chrome PDF settings", async () => {
    await goto("chrome://settings/content/pdfDocuments",1000);
    await waitFor(2000);
    await click($('#bar'))
    await waitFor(2000);
});

step("Goto site", async () => {
await goto(url,3000);   
});

step("Search <a> with <b> and <c>", async (a, b, c) => {

try{
    await clear(textBox("A"));
    await write(a, into(textBox("A")))
    await click(button("Submit Search"));
    
	try
    {   await focus($("#pdf_" + a + "_" + b));
        await click($("#pdf_" + a + "_" + b));
        await waitFor('2000');
	}
    catch{}
}
catch{}	
```
 The search step uses a table and the error occurs after `await click($("#pdf_" + a + "_" + b));` and before `await waitFor('2000');`

in the catches I log and recover from some errors.

**Reply by NivedhaSenthil** [reply]
Thanks @KarianneM for the note... it would be difficult to analyze the root cause without a replica but I will spend some time analyzing through the stacktrace

**Comment by KarianneM** [comment]
I switched to headless and the error stopped.

**Reply by sankarasai18-gauge** [reply]
@KarianneM @NivedhaSenthil @zabil This is the same issue which I am facing as posted in this thread (https://github.com/getgauge/gauge/issues/2165) . Even for me with headless mode on I am gettting this issue (Runner is not alive).
Please help me to fix this issue.
